### ISSUE TITLE
[java] New Rule: AvoidReassigningCatchVariables

### ISSUE THREAD

**Issue by numeralnathan** [pmd/pmd#2471]
**Proposed Rule Name:** AvoidReassigningCatchVariables

**Proposed Category:** Best Practices

**Description:**

The rule [AvoidReassigningParameters](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#avoidreassigningparameters) flags a problem when code assigns a value to a method parameter.  Here is the reason in documentation.

> Reassigning values to incoming parameters is not recommended. Use temporary local variables instead.

This new rule `AvoidReassigningCatchVariables` will flag a problem when code assigns a value to the catch variable.

I do not know why `AvoidReassigningParameters` exists.  I assume that `AvoidReassigningCatchVariables` should exist for the same reason.

**Code Sample:**

```java
...
catch (Exception e)
{
   e = new Exception("Hello World", e);   // This should be flagged as a problem
}
...

// Fixed code
...
catch (Exception e)
{
   Exception wrap;

   wrap = new Exception("Hello World", e);
}
```

**Possible Properties:**

This rule does not need any properties.

**See also:**

* #2219

**Comment by oowekyala** [comment]
One justification this rule could have, that is more solid than that of AvoidReassigningParameters, is that exception parameters in a multicatch block are implicitly final. So if you do
```java
catch (IOException e)
{
   e = new Exception("Hello World", e); 
}
```
adding an exception type will make it uncompilable:
```java
catch (IOException | TokenMgrError e)
{
   e = new Exception("Hello World", e); // e is implicitly final, this doesn't compile
}
```

**Comment by adangel** [comment]
On the other hand - if the compiler already complains, then we don't need a rule :smile: - Update - I think, I get it now: the rule would prevent this compile error by flagging it even before (when assignment would be possible), so that you actually *can* add later another exception to the case and make it a multicatch.

Reassigning parameters or catch variables *might* be surprising. It depends on what you are used to (principle of least surprise). If you assume, that the parameters/catch vars always have their initial value when entering the block, then you might make false decisions, if this is not the case (e.g. it might not be anymore the original exception with the correct/original stacktrace). See also #2219.
